File: src/sql/shared/ObjectIterator.zig
const ObjectIterator = @This();

array: JSValue,
columns: JSValue = .zero,
globalObject: *jsc.JSGlobalObject,
cell_i: usize = 0,
row_i: usize = 0,
current_row: jsc.JSValue = .zero,
columns_count: usize = 0,
array_length: usize = 0,
any_failed: bool = false,

pub fn next(this: *ObjectIterator) ?jsc.JSValue {
    if (this.array.isEmptyOrUndefinedOrNull() or this.columns.isEmptyOrUndefinedOrNull()) {
        this.any_failed = true;
        return null;
    }
    if (this.row_i >= this.array_length) {
        return null;
    }

    const cell_i = this.cell_i;
    this.cell_i += 1;
    const row_i = this.row_i;

    const globalObject = this.globalObject;

    if (this.current_row == .zero) {
        this.current_row = jsc.JSObject.getIndex(this.array, globalObject, @intCast(row_i)) catch {
            this.any_failed = true;
            return null;
        };
        if (this.current_row.isEmptyOrUndefinedOrNull()) {
            return globalObject.throw("Expected a row to be returned at index {d}", .{row_i}) catch null;
        }
    }

    defer {
        if (this.cell_i >= this.columns_count) {
            this.cell_i = 0;
            this.current_row = .zero;
            this.row_i += 1;
        }
    }

    const property = jsc.JSObject.getIndex(this.columns, globalObject, @intCast(cell_i)) catch {
        this.any_failed = true;
        return null;
    };
    if (property.isUndefined()) {
        return globalObject.throw("Expected a column at index {d} in row {d}", .{ cell_i, row_i }) catch null;
    }

    const value = this.current_row.getOwnByValue(globalObject, property);
    if (value == .zero or (value != null and value.?.isUndefined())) {
        if (!globalObject.hasException())
            return globalObject.throw("Expected a value at index {d} in row {d}", .{ cell_i, row_i }) catch null;
        this.any_failed = true;
        return null;
    }
    return value;
}

const bun = @import("bun");

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/shared/ObjectIterator.rs
use crate::jsc::{JSGlobalObject, JSObject, JSValue, JsResult};

// Note: this iterator holds bare `JSValue` fields and a borrowed
// `&JSGlobalObject`; it is only sound when constructed on the stack for the
// duration of a single bind/iteration pass (conservative GC stack scan keeps
// `array`/`columns`/`current_row` alive). Never `Box` this.
pub(crate) struct ObjectIterator<'a> {
    pub(crate) array: JSValue,
    pub(crate) columns: JSValue,
    pub(crate) global_object: &'a JSGlobalObject,
    pub(crate) cell_i: u32,
    pub(crate) row_i: u32,
    pub(crate) current_row: JSValue,
    pub(crate) columns_count: u32,
    pub(crate) array_length: u32,
}

impl<'a> ObjectIterator<'a> {
    /// The next cell value (row `row_i`, column `cell_i`), `Ok(None)` once every row has been visited.
    pub(crate) fn next(&mut self) -> JsResult<Option<JSValue>> {
        if self.row_i >= self.array_length {
            return Ok(None);
        }

        let cell_i = self.cell_i;
        self.cell_i += 1;
        let row_i = self.row_i;

        let global_object = self.global_object;

        if self.current_row.is_empty() {
            self.current_row = JSObject::get_index(self.array, global_object, row_i)?;
            if !self.current_row.is_object() {
                return Err(
                    global_object.throw(format_args!("Expected a row object at index {}", row_i))
                );
            }
        }

        let property = JSObject::get_index(self.columns, global_object, cell_i)?;
        if property.is_undefined() {
            return Err(global_object.throw(format_args!(
                "Expected a column at index {} in row {}",
                cell_i, row_i
            )));
        }
        let value = self.current_row.get_own_by_value(global_object, property)?;
        let result = if value.is_undefined() {
            Err(global_object.throw(format_args!(
                "Expected a value at index {} in row {}",
                cell_i, row_i
            )))
        } else {
            Ok(Some(value))
        };

        if self.cell_i >= self.columns_count {
            self.cell_i = 0;
            self.current_row = JSValue::ZERO;
            self.row_i += 1;
        }

        result
    }
}
